pyro/infer/mcmc/mcmc.py

```diff
@@ -278,8 +251,61 @@ def __init__(self, kernel, num_samples, warmup_steps=0,
                                             num_chains, mp_context)
         else:
             self.sampler = _SingleSampler(kernel, num_samples, warmup_steps)
-        super(MCMC, self).__init__()
 
     def _traces(self, *args, **kwargs):
         for sample in self.sampler._traces(*args, **kwargs):
             yield sample
+
+    def marginal(self, sites=None):
+        return _EmpiricalMarginal(self, sites)
+
+
+class _EmpiricalMarginal(object):
+    """
+    Marginal distribution for MCMC that provides a a marginal over one or more
+    latent sites as well as the return values of the TracePosterior's model.
+
+    :param TracePosterior trace_posterior: a TracePosterior instance representing
+        a Monte Carlo posterior.
+    :param list sites: optional list of sites for which we need to generate
+        the marginal distribution. Note that for multiple sites, the shape
+        for the site values must match (needed by the underlying ``Empirical``
+        class).
+    """
+    def __init__(self, trace_posterior, sites=None, validate_args=None):
+        assert isinstance(trace_posterior, TracePosterior), \
+            "trace_dist must be trace posterior distribution object"
+        if sites is None:
+            sites = ["_RETURN"]
+        elif isinstance(sites, str):
+            sites = [sites]
+        else:
+            assert isinstance(sites, list)
+        self.sites = sites
+        self._marginals = OrderedDict()
+        self._diagnostics = OrderedDict()
+        self._populate_traces(trace_posterior, validate_args)
+
+    def _populate_traces(self, trace_posterior, validate):
+        with ExitStack() as stack:
+            self._marginals = {site: stack.enter_context(accumulate_samples(validate))
+                               for site in self.sites}
+            for tr, log_weight, chain_id in zip(trace_posterior.exec_traces,
+                                                trace_posterior.log_weights,
+                                                trace_posterior.chain_ids):
+                for site in self._marginals:
+                    value = tr.nodes[site]["value"]
+                    self._marginals[site].add(value, log_weight=log_weight, chain_id=chain_id)
+
+    def diagnostics(self):
+        if self._diagnostics:
+            return self._diagnostics
+        for site in self.sites:
+            self._diagnostics[site] = OrderedDict([
+                ("n_eff", stats.effective_sample_size(self._marginals[site].get_data())),
+                ("r_hat", stats.split_gelman_rubin(self._marginals[site].get_data()))
+            ])
+        return self._diagnostics
+
+    def empirical(self):
+        return self._marginals
```

Comment:
I think we can make this a property.

Reply:
Or we can make EmpiricalMarginal a dict-like object. If you agree, I'll do that in another pull request.

Reply:
Hmm..I'm not sure if that will work. Can you elaborate?

Reply:
This is what I have in mind (not sure it is indeed useful at all):

```
def items(self):
    for name in self._marginals:
        yield name, self[name]

def keys(self):
    return self._marginals.keys()

def values(self):
    for name, empirical in self.items():
        yield empirical

def __bool__(self):
    return bool(self._marginals)

def __len__(self):
    return len(self._marginals)

def __contains__(self, name):
    return name in self._marginals

def __iter__(self):
    return iter(self.keys())

def __delitem__(self, name):
    self.sites.remove(value)
    del self._marginals[name]
    if self._diagnostics:
        del self._diagnostics[name]

def __getitem__(self, name):
    return self._marginals[name]
    
def __setitem__(self, name, new_empirical):
    if name not in self.sites:
        self.sites.append(name)
    self._marginals[name] = new_empirical
    if self._diagnostics:
         self._diagnostics[name] = OrderedDict([...])
```

Reply:
We have changed the API as follows:

Earlier:
```python
mcmc_run = MCMC(hmc_kernel, ..).run(data)
marginal = EmpiricalMarginal(mcmc_run, sites=["a", "b"]).get_dist_and_values()[0]
# this will only work if a and b had the same shape
# version below does not have this issue.
samples_a = marginal[:, 1]
samples_b = marginal[:, 2]
```

Now:
```python
mcmc_run = MCMC(hmc_kernel, ..).run(data)
marginal = mcmc_run.marginal(["a", "b"])
samples_a = mcmc_run.empirical["a"].enumerate_support()
samples_b = mcmc_run.empirical["b"].enumerate_support()
```

The second version has an additional indirection, because `.marginal()` is not a distribution unlike `EmpiricalMarginal`, and we need to call `.empirical` to get the underlying empirical distribution. We can provide a convenience method called `.support` which can do this under the hood though:

```python
mcmc_run = MCMC(hmc_kernel, ..).run(data)
marginal = mcmc_run.marginal(["a", "b"]).support
samples_a = marginal["a"]
samples_b = marginal["b"]
```

I don't completely follow how the dict interface will simplify this.

Reply:
@neerajprad I think that
```
mcmc_run = MCMC(hmc_kernel, ..).run(data)
marginal = mcmc_run.marginal(["a", "b"])
samples_a = marginal.empirical["a"].enumerate_support()
samples_b = marginal.empirical["b"].enumerate_support()
```
is pretty clean! Please don't worry about `dict`.